<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.301">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>EEB313: Quantitative Methods in R for Biology - 9&nbsp; Linear models</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./projects.html" rel="next">
<link href="./lec07-intro-to-stats.html" rel="prev">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script>

  <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

</head>

<body class="nav-sidebar floating">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
      <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./lec00-rstudio.html">Lectures</a></li><li class="breadcrumb-item"><a href="./lec08-linear-models.html"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Linear models</span></a></li></ol></nav>
      <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
      </a>
      <button type="button" class="btn quarto-search-button" aria-label="Search" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">EEB313: Quantitative Methods in R for Biology</a> 
        <div class="sidebar-tools-main">
    <a href="https://github.com/EEB313/eeb313.github.io" title="Source Code" class="quarto-navigation-tool px-1" aria-label="Source Code"><i class="bi bi-github"></i></a>
</div>
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Syllabus</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./about-us.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">The 2023 teaching team</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./downloadingR.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Downloading R</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">
 <span class="menu-text">Lectures</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./lec00-rstudio.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Getting started with RStudio and R Notebook</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./lec01-markdown-workflows.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Markdown, project workflows</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./lec02-base-r.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Base R: assignment, vectors, functions, and loops</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./lec03-dataframes-dplyr.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Data frames, intro to dplyr</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./lec04-data-wrangling.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Data wrangling in dplyr</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./lec05-data-visualization.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Data visualization in ggplot</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./lec06-exploratory-data-analysis.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Exploratory data analysis</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./lec07-intro-to-stats.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Introduction to inference</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./lec08-linear-models.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Linear models</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <span class="sidebar-item-text sidebar-link text-start">
 <span class="menu-text">Assignments</span></span>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true">
 <span class="menu-text">Project</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./projects.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Project description</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./databases.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">Some open-access databases</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="true">
 <span class="menu-text">Miscellaneous notes</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./tests-assignment-3.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">More on hypothesis testing</span></span></a>
  </div>
</li>
      </ul>
  </li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#lesson-preamble" id="toc-lesson-preamble" class="nav-link active" data-scroll-target="#lesson-preamble"><span class="header-section-number">9.1</span> Lesson preamble:</a></li>
  <li><a href="#linear-models-why-we-care" id="toc-linear-models-why-we-care" class="nav-link" data-scroll-target="#linear-models-why-we-care"><span class="header-section-number">9.2</span> Linear models: why we care</a></li>
  <li><a href="#theory-likelihood-estimation-and-inference-for-linear-models" id="toc-theory-likelihood-estimation-and-inference-for-linear-models" class="nav-link" data-scroll-target="#theory-likelihood-estimation-and-inference-for-linear-models"><span class="header-section-number">9.3</span> Theory: likelihood estimation and inference for linear models</a>
  <ul class="collapse">
  <li><a href="#simple-linear-regression-p1" id="toc-simple-linear-regression-p1" class="nav-link" data-scroll-target="#simple-linear-regression-p1"><span class="header-section-number">9.3.1</span> Simple linear regression: <span class="math inline">\(p=1\)</span></a></li>
  <li><a href="#the-matrix-vesion-of-the-general-linear-model" id="toc-the-matrix-vesion-of-the-general-linear-model" class="nav-link" data-scroll-target="#the-matrix-vesion-of-the-general-linear-model"><span class="header-section-number">9.3.2</span> The matrix vesion of the general linear model</a></li>
  </ul></li>
  <li><a href="#practice-fitting-linear-models-with-lm" id="toc-practice-fitting-linear-models-with-lm" class="nav-link" data-scroll-target="#practice-fitting-linear-models-with-lm"><span class="header-section-number">9.4</span> Practice: fitting linear models with <code>lm</code></a>
  <ul class="collapse">
  <li><a href="#transformations" id="toc-transformations" class="nav-link" data-scroll-target="#transformations"><span class="header-section-number">9.4.1</span> Transformations</a></li>
  <li><a href="#syntax-for-multiple-regression-interactions" id="toc-syntax-for-multiple-regression-interactions" class="nav-link" data-scroll-target="#syntax-for-multiple-regression-interactions"><span class="header-section-number">9.4.2</span> Syntax for multiple regression, interactions</a></li>
  <li><a href="#the-relationship-between-linear-models-and-anova" id="toc-the-relationship-between-linear-models-and-anova" class="nav-link" data-scroll-target="#the-relationship-between-linear-models-and-anova"><span class="header-section-number">9.4.3</span> The relationship between linear models and ANOVA</a></li>
  <li><a href="#a-fun-application" id="toc-a-fun-application" class="nav-link" data-scroll-target="#a-fun-application"><span class="header-section-number">9.4.4</span> A fun application!</a></li>
  </ul></li>
  <li><a href="#generalized-linear-models-theory-and-examples" id="toc-generalized-linear-models-theory-and-examples" class="nav-link" data-scroll-target="#generalized-linear-models-theory-and-examples"><span class="header-section-number">9.5</span> Generalized linear models: theory and examples</a>
  <ul class="collapse">
  <li><a href="#returning-to-the-disease-age-example" id="toc-returning-to-the-disease-age-example" class="nav-link" data-scroll-target="#returning-to-the-disease-age-example"><span class="header-section-number">9.5.1</span> Returning to the disease-age example…</a></li>
  </ul></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title"><span id="lec08-linear-models" class="quarto-section-identifier"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Linear models</span></span></h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header>

<section id="lesson-preamble" class="level2" data-number="9.1">
<h2 data-number="9.1" class="anchored" data-anchor-id="lesson-preamble"><span class="header-section-number">9.1</span> Lesson preamble:</h2>
<blockquote class="blockquote">
<h3 id="lesson-objectives" data-number="9.1.1" class="anchored"><span class="header-section-number">9.1.1</span> Lesson objectives</h3>
<ul>
<li>Understand the logic of simple and multiple linear regression, including the assumptions that are placed on the data, parameters, and errors.</li>
<li>Understand the meaning of regression coefficients and how they are estimated.</li>
<li>Learn how confidence intervals and <span class="math inline">\(p\)</span>-values associated to the regression coefficients are calculated and used to test hypotheses.</li>
<li>Understand how to implement linear models (including ANOVA) in R.</li>
<li>Develop familiarity with generalized linear models and some important examples (logistic, Poisson, negative binomial regression).</li>
</ul>
<h3 id="lesson-outline" data-number="9.1.2" class="anchored"><span class="header-section-number">9.1.2</span> Lesson outline</h3>
<ul>
<li>Linear models: theory and examples
<ul>
<li>Structure and assumptions, including interpretation of the effects</li>
<li>Likelihood-based estimation and inference</li>
<li>Transformations</li>
<li>Dummy variables, interactions between covariates, etc.</li>
<li>Analysis of Variance</li>
</ul></li>
<li>Generalized linear models
<ul>
<li>Non-normal errors, link functions</li>
<li>Estimation and inference: even more likelihood!</li>
<li>Logistic regression</li>
<li>Poisson, negative binomial regression</li>
</ul></li>
</ul>
</blockquote>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(car)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a>surveys_subset <span class="ot">&lt;-</span> <span class="fu">read_csv</span>(<span class="st">"survey_subset.csv"</span>)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="fu">theme_set</span>(<span class="fu">theme_bw</span>())</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="linear-models-why-we-care" class="level2" data-number="9.2">
<h2 data-number="9.2" class="anchored" data-anchor-id="linear-models-why-we-care"><span class="header-section-number">9.2</span> Linear models: why we care</h2>
<p>Linear models are at the heart of statistical practice in the physical, life, and social sciences! Linear regression actually refers to a <em>family</em> of modeling approaches that attempt to learn how the mean and/or variance of a response variable <span class="math inline">\(\boldsymbol{y} = (y_1,\dots,y_n)\)</span> depend on (linear) combinations of variables <span class="math inline">\(\boldsymbol{x}_i = (x_{i1},\dots,x_{in})\)</span> called predictors. In this lecture, we will discuss various forms of the linear model and assumptions placed on the data to make estimation and inference of the relationships between variables tractable. We will see how the likelihood function forms the basis for this estimation/inference, and how extensions of multiple regression (generalized linear models and mixed models!) can be understood in a likelihood framework. Our goal will be to become familiar with how these models work and how they are fit to data in R.</p>
</section>
<section id="theory-likelihood-estimation-and-inference-for-linear-models" class="level2" data-number="9.3">
<h2 data-number="9.3" class="anchored" data-anchor-id="theory-likelihood-estimation-and-inference-for-linear-models"><span class="header-section-number">9.3</span> Theory: likelihood estimation and inference for linear models</h2>
<p>A linear model takes the form</p>
<p><span class="math display">\[ y_i = \beta_0 + \beta_1 x_{1i} + \cdots + \beta_p x_{pi} + \varepsilon_{i}, \]</span></p>
<p>where <span class="math inline">\(i=1,\dots,n\)</span> correspond to observations of a random variable <span class="math inline">\(Y\)</span> which we call the <em>response</em>. <strong>The goal of regression is to explain and to predict the behavior of the response variable by adding together effects of <span class="math inline">\(p\)</span> covariates <span class="math inline">\(x_1,\dots,x_p\)</span>, which are also called <em>predictors</em>.</strong> Importantly, the phrase “linear model” is somewhat deceptive in that <strong>the model can be used to describe many kinds of functional relationships in the data</strong>: for example, we could use <span class="math inline">\(x_2 = x_1^2\)</span> to model higher order effects of <span class="math inline">\(x_1\)</span> on the response. In general, the linear model above specifies how the realizations <span class="math inline">\(y_1,\dots,y_n\)</span> of a random variable <span class="math inline">\(Y\)</span> depend on the additive effects of one or more non-random covariates/predictors. We will discuss random and mixed effects next class, but the machinery used to estimate the effect sizes <span class="math inline">\(\beta_1,\dots,\beta_p\)</span> and error variance are very similar to the model in which the <span class="math inline">\(x\)</span>s are assumed to be fixed.</p>
<p>To make estimation and inference tractable, the errors <span class="math inline">\(\varepsilon_i\)</span> are assumed to be Normal with mean zero and variance <span class="math inline">\(\sigma_i^2\)</span>. Equivalently, we could assume the <em>data</em> <span class="math inline">\(y_1,\dots,y_n\)</span> are 1) independent and 2) Normal with mean(s) <span class="math inline">\(\beta_0 + \beta_1 x_{1i} + \cdots + \beta_p x_{pi}\)</span> and variance(s) <span class="math inline">\(\sigma_1^2,\dots,\sigma_n^2\)</span>. Although the normality assumption is not strictly necessary (nor is the assumption of equal variance across observations), they are commonly made and allow us to easily estimate the effect sizes and error variance. Under the assumption of normality and constant error variance <span class="math inline">\(\sigma_1^2 = \dots = \sigma_n^2 = \sigma^2\)</span>, the likelihood can be written as</p>
<p><span class="math display">\[ L(\beta_0,\beta_1,\dots,\beta_p,\sigma^2) = \prod_{i=1}^n (2 \pi \sigma^2)^{-1/2} e^{-(y_i - \beta_0 - \beta_1 x_{1i} - \dots - \beta_p x_{pi})^2/2\sigma^2}. \]</span></p>
<p>Estimators of the regression coefficients <span class="math inline">\(\beta_1,\dots,\beta_p\)</span> and <span class="math inline">\(\sigma^2\)</span> are found by maximizing the likelihood. In fact, the <strong>estimators found by maximum likelihood methods are exactly those which minimize the distance from the model <span class="math inline">\(y = \beta_0 + \beta_1 x_{1} + \dots + \beta_p x_{p}\)</span> from the data</strong>. The line (and in higher dimensions, plane) given by this equation is such that the the sum of squared departures of the line from the data is as small is it can be. Other notions of distance turn out to give rise to interesting extensions of the linear model, such as ridge and LASSO regression, which are beyond the scope of the course.</p>
<section id="simple-linear-regression-p1" class="level3" data-number="9.3.1">
<h3 data-number="9.3.1" class="anchored" data-anchor-id="simple-linear-regression-p1"><span class="header-section-number">9.3.1</span> Simple linear regression: <span class="math inline">\(p=1\)</span></h3>
<p>Simple linear regression is a special case of the general linear model (above), and corresponds to the case there is only one predictor, i.e., <span class="math inline">\(p=1\)</span>.</p>
</section>
<section id="the-matrix-vesion-of-the-general-linear-model" class="level3" data-number="9.3.2">
<h3 data-number="9.3.2" class="anchored" data-anchor-id="the-matrix-vesion-of-the-general-linear-model"><span class="header-section-number">9.3.2</span> The matrix vesion of the general linear model</h3>
<p>What is often used in theory and practice is a matrix version of the above model:</p>
<p><span class="math display">\[ \boldsymbol{y} = \begin{bmatrix}
     y_{1} \\
     y_{2} \\
     \vdots \\
     y_{n}
     \end{bmatrix} = \begin{bmatrix}
     1 &amp; x_{11} &amp; x_{21} &amp; \cdots &amp; x_{p1} \\
     1 &amp; x_{12} &amp; x_{22} &amp; \cdots &amp; x_{p2} \\
     \vdots &amp; \vdots &amp; \vdots &amp; &amp; \vdots \\
     1 &amp; x_{1n} &amp; x_{2n} &amp; \cdots &amp; x_{pn}
     \end{bmatrix} \begin{bmatrix}
     \beta_{1} \\
     \beta_{2} \\
     \vdots \\
     \beta_{p}
     \end{bmatrix} + \begin{bmatrix}
     \varepsilon_{1} \\
     \varepsilon_{2} \\
     \vdots \\
     \varepsilon_{n}
     \end{bmatrix} = \boldsymbol{X} \boldsymbol{\beta} + \boldsymbol{\varepsilon}.
  \]</span></p>
<p>Here, <span class="math inline">\(\boldsymbol{y} = (y_1,\dots,y_n)'\)</span> is a vector of measurements for the response, <span class="math inline">\(\boldsymbol{x_i} = (x_{i1},\dots,x_{in})'\)</span> is a vector of measurements for the <span class="math inline">\(k\)</span>th predictor, and <span class="math inline">\(\boldsymbol{\varepsilon} = (\varepsilon_1,\dots,\varepsilon_n)'\)</span> is a vector of measurement errors. The <span class="math inline">\('\)</span> symbol corresponds to transposition, which involves interchanging the rows and columns of a vector or matrix.<a href="#fn1" class="footnote-ref" id="fnref1" role="doc-noteref"><sup>1</sup></a></p>
<p>When all the errors are modeled as above (i.e., as Normal with constant variance and mean zero), the ML estimator for the vector of regression coefficients is <span class="math inline">\(\boldsymbol{\hat{\beta}_{\text{MLE}}} = (\boldsymbol{X'X})^{-1} \boldsymbol{X'} \boldsymbol{y}\)</span>.</p>
<section id="what-do-the-regression-coefficients-mean" class="level4" data-number="9.3.2.1">
<h4 data-number="9.3.2.1" class="anchored" data-anchor-id="what-do-the-regression-coefficients-mean"><span class="header-section-number">9.3.2.1</span> What do the regression coefficients mean?</h4>
<p>The interpretation of the regression coefficients is as follows: <strong><span class="math inline">\(\beta_j\)</span> describes how much the response is expected to change, <em>all else constant</em>, if we increase <span class="math inline">\(x_j\)</span> by exactly one unit.</strong> Importantly, the <em>fitted</em> regression coefficients measure the effect (slope) of increasing <span class="math inline">\(x_1,\dots,x_j\)</span> on the response <em>under the assumption of normal data</em> — this distinction between the theoretical and fitted coefficients is important to keep mind.</p>
</section>
<section id="categorical-predictors" class="level4" data-number="9.3.2.2">
<h4 data-number="9.3.2.2" class="anchored" data-anchor-id="categorical-predictors"><span class="header-section-number">9.3.2.2</span> Categorical predictors</h4>
<p>Before we dive into implementing linear models in R, it is important to mention how the linear model accommodates discrete predictors like sex (or genotype, ID, race). To deal with categorical predictors, we define the model in terms of a <em>baseline</em> and to interpret the regression coefficients <em>relative</em> to this baseline. This involves coding “dummy variables” <span class="math inline">\(x_1,\dots,x_{k-1}\)</span> for all but one the values (<span class="math inline">\(1,2,\dots,k\)</span>) the predictor can take one, so that <span class="math inline">\(x_{ji} = 1\)</span> for observations where the categorical variable is <span class="math inline">\(=j\)</span> and <span class="math inline">\(=0\)</span> otherwise.</p>
</section>
</section>
</section>
<section id="practice-fitting-linear-models-with-lm" class="level2" data-number="9.4">
<h2 data-number="9.4" class="anchored" data-anchor-id="practice-fitting-linear-models-with-lm"><span class="header-section-number">9.4</span> Practice: fitting linear models with <code>lm</code></h2>
<p>To illustrate how regression models are fitted (via maximum likelihood) in R, we will use the survey dataset from a couple classes ago. We begin by regressing weight on hindfoot length:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>model <span class="ot">&lt;-</span> <span class="fu">lm</span>(weight <span class="sc">~</span> hindfoot_length, <span class="at">data=</span>surveys_subset)</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(model) <span class="co"># assuming normality, homogeneity of variance</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Call:
lm(formula = weight ~ hindfoot_length, data = surveys_subset)

Residuals:
    Min      1Q  Median      3Q     Max 
-60.342 -13.126  -4.018   3.351 221.135 

Coefficients:
                Estimate Std. Error t value Pr(&gt;|t|)    
(Intercept)     -33.2179     1.3575  -24.47   &lt;2e-16 ***
hindfoot_length   2.5651     0.0441   58.16   &lt;2e-16 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 26.19 on 3842 degrees of freedom
  (505 observations deleted due to missingness)
Multiple R-squared:  0.4682,    Adjusted R-squared:  0.4681 
F-statistic:  3383 on 1 and 3842 DF,  p-value: &lt; 2.2e-16</code></pre>
</div>
</div>
<p>R returns the following after fitting a linear model via <code>lm()</code>:</p>
<ul>
<li><strong>Descriptive statistics for the “residuals”</strong> <span class="math inline">\(\varepsilon_i = y_i - \widehat{\beta_0} - \widehat{\beta_1} x_i\)</span>, which tell us about how much variability there is in the data relative to the linear model specified and fitted.</li>
<li>The <strong>regression coefficients minimizing the sum of squared departures from the data (i.e., the ML estimates) and <span class="math inline">\(95\%\)</span> confidence intervals for each</strong>. The CIs are expressed as standard errors, since the estimators have an approximate Normal distribution. A <em>joint</em> confidence region for the coefficients can also be found using, e.g., the LRT statistic.</li>
<li>A <strong>suite of test statistics</strong>! The <span class="math inline">\(t\)</span> statistics and their <span class="math inline">\(p\)</span> values are associated to the test <span class="math inline">\(H_0: \beta_i = 0\)</span> vs <span class="math inline">\(H_1: \beta_i \neq 0\)</span>. Significance codes specify the level <span class="math inline">\(\alpha\)</span> at which we have evidence to reject the null hypothesis for each coefficient.</li>
<li>Measures of goodness-of-fit: <strong>the multiple <span class="math inline">\(R^2\)</span> and the adjusted <span class="math inline">\(R^2\)</span></strong>. These explain the proportion of variance that are explained by the model. The latter measures the proportion of variance explained by the linear model upon adjusting for sample size and <span class="math inline">\(\#\)</span> of predictors.</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(surveys_subset, <span class="fu">aes</span>(<span class="at">x =</span> hindfoot_length, <span class="at">y =</span> weight)) <span class="sc">+</span> </span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">color =</span> <span class="fu">as.factor</span>(sex)), <span class="at">size =</span> <span class="dv">2</span>) <span class="sc">+</span></span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">stat_smooth</span>(<span class="at">method =</span> <span class="st">"lm"</span>, <span class="at">se =</span> F, <span class="at">color =</span> <span class="st">"gray"</span>) <span class="sc">+</span> <span class="fu">labs</span>(<span class="at">color =</span> <span class="st">"sex"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>`geom_smooth()` using formula = 'y ~ x'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Removed 505 rows containing non-finite values (`stat_smooth()`).</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Removed 505 rows containing missing values (`geom_point()`).</code></pre>
</div>
<div class="cell-output-display">
<p><img src="lec08-linear-models_files/figure-html/checkIfDataAreNormal-1.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="do">### are the data normal?</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(surveys_subset, <span class="fu">aes</span>(<span class="at">x =</span> weight)) <span class="sc">+</span> <span class="fu">geom_histogram</span>() <span class="do">## oh no!</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>`stat_bin()` using `bins = 30`. Pick better value with `binwidth`.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Removed 316 rows containing non-finite values (`stat_bin()`).</code></pre>
</div>
<div class="cell-output-display">
<p><img src="lec08-linear-models_files/figure-html/checkIfDataAreNormal-2.png" class="img-fluid" width="672"></p>
</div>
</div>
<section id="transformations" class="level3" data-number="9.4.1">
<h3 data-number="9.4.1" class="anchored" data-anchor-id="transformations"><span class="header-section-number">9.4.1</span> Transformations</h3>
<p>Often, data are non-normal! This is an unfortunate fact of life. It is sometimes possible, however, to use the machinery of regression if there is a suitable <em>transformation</em> of the data which makes it normal, e.g., <code>log()</code>, <code>sqrt()</code>. Right-skewed data (like above) may be normalized using log or root transformations (e.g.&nbsp;square root, third-root, etc.), with greater roots required for increasingly right-skewed data. Left-skewed data could be normalized with power transformations (e.g.&nbsp;squared, 3rd power, etc.).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(surveys_subset, <span class="fu">aes</span>(<span class="at">x =</span> <span class="fu">log</span>(weight))) <span class="sc">+</span> <span class="fu">geom_histogram</span>() <span class="do">## better!</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>`stat_bin()` using `bins = 30`. Pick better value with `binwidth`.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Removed 316 rows containing non-finite values (`stat_bin()`).</code></pre>
</div>
<div class="cell-output-display">
<p><img src="lec08-linear-models_files/figure-html/unnamed-chunk-2-1.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>model_logtrasnformed <span class="ot">&lt;-</span> <span class="fu">lm</span>(<span class="fu">log</span>(weight)<span class="sc">~</span>hindfoot_length, <span class="at">data=</span>surveys_subset)</span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(model_logtrasnformed)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Call:
lm(formula = log(weight) ~ hindfoot_length, data = surveys_subset)

Residuals:
    Min      1Q  Median      3Q     Max 
-1.8285 -0.2135 -0.0473  0.1488  1.9612 

Coefficients:
                 Estimate Std. Error t value Pr(&gt;|t|)    
(Intercept)     1.5655038  0.0203235   77.03   &lt;2e-16 ***
hindfoot_length 0.0647424  0.0006603   98.06   &lt;2e-16 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 0.3922 on 3842 degrees of freedom
  (505 observations deleted due to missingness)
Multiple R-squared:  0.7145,    Adjusted R-squared:  0.7144 
F-statistic:  9615 on 1 and 3842 DF,  p-value: &lt; 2.2e-16</code></pre>
</div>
</div>
<p>As before, there is evidence to reject the null hypothesis that hindfoot length has no effect on weight (or, in this case, its log transformation) at significance level <span class="math inline">\(\alpha = 0.05\)</span>. This is because the <span class="math inline">\(p\)</span>-value that is associated to the coefficent of hindfoot length is <span class="math inline">\(&lt; \alpha\)</span>. We can also use what is returned by <code>lm()</code> to predict what the response will be if we observe new data (hindfoot lengths).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>new_hindfoot_length_obs <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">hindfoot_length =</span> <span class="fu">seq</span>(<span class="dv">0</span>,<span class="dv">100</span>,<span class="fl">0.1</span>))</span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>predicted_values <span class="ot">&lt;-</span> <span class="fu">predict.lm</span>(<span class="at">object =</span> model_logtrasnformed, </span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a>                               <span class="at">newdata =</span> new_hindfoot_length_obs)</span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="fu">cbind</span>(<span class="at">logweight =</span> predicted_values, new_hindfoot_length_obs), </span>
<span id="cb16-6"><a href="#cb16-6" aria-hidden="true" tabindex="-1"></a>       <span class="fu">aes</span>(<span class="at">x =</span> hindfoot_length, <span class="at">y =</span> logweight)) <span class="sc">+</span> <span class="fu">geom_line</span>(<span class="at">size =</span> <span class="dv">1</span>, <span class="at">color =</span> <span class="st">"gray"</span>) <span class="sc">+</span> </span>
<span id="cb16-7"><a href="#cb16-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">data =</span> surveys_subset, <span class="at">inherit.aes =</span> F,  <span class="at">size =</span> <span class="dv">2</span>, </span>
<span id="cb16-8"><a href="#cb16-8" aria-hidden="true" tabindex="-1"></a>             <span class="fu">aes</span>(<span class="at">x =</span> hindfoot_length, <span class="at">y =</span> <span class="fu">log</span>(weight), <span class="at">color =</span> <span class="fu">as.factor</span>(sex))) <span class="sc">+</span> </span>
<span id="cb16-9"><a href="#cb16-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">color =</span> <span class="st">"sex"</span>, <span class="at">y =</span> <span class="st">"log(weight)"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Using `size` aesthetic for lines was deprecated in ggplot2 3.4.0.
ℹ Please use `linewidth` instead.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Removed 505 rows containing missing values (`geom_point()`).</code></pre>
</div>
<div class="cell-output-display">
<p><img src="lec08-linear-models_files/figure-html/unnamed-chunk-3-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<section id="challenge" class="level4" data-number="9.4.1.1">
<h4 data-number="9.4.1.1" class="anchored" data-anchor-id="challenge"><span class="header-section-number">9.4.1.1</span> Challenge</h4>
<p>Regress weight on sex. Be sure to specify that sex is a factor in the call to <code>lm()</code>. What value of sex is used as a baseline? Is there a significant effect of sex on weight?</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>model <span class="ot">&lt;-</span> <span class="fu">lm</span>(weight<span class="sc">~</span><span class="fu">as.factor</span>(sex), <span class="at">data=</span>surveys_subset)</span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(model) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Call:
lm(formula = weight ~ as.factor(sex), data = surveys_subset)

Residuals:
   Min     1Q Median     3Q    Max 
-38.14 -22.14  -6.99   5.01 227.86 

Coefficients:
                Estimate Std. Error t value Pr(&gt;|t|)    
(Intercept)      42.9896     0.8403  51.160   &lt;2e-16 ***
as.factor(sex)M  -0.8537     1.1651  -0.733    0.464    
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 36.89 on 4014 degrees of freedom
  (333 observations deleted due to missingness)
Multiple R-squared:  0.0001337, Adjusted R-squared:  -0.0001154 
F-statistic: 0.5368 on 1 and 4014 DF,  p-value: 0.4638</code></pre>
</div>
</div>
</section>
</section>
<section id="syntax-for-multiple-regression-interactions" class="level3" data-number="9.4.2">
<h3 data-number="9.4.2" class="anchored" data-anchor-id="syntax-for-multiple-regression-interactions"><span class="header-section-number">9.4.2</span> Syntax for multiple regression, interactions</h3>
<p>One can regress on several explanatory variables simultaneously as follows:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>model <span class="ot">&lt;-</span> <span class="fu">lm</span>(weight<span class="sc">~</span>hindfoot_length<span class="sc">+</span><span class="fu">as.factor</span>(sex), <span class="at">data=</span>surveys_subset)</span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(model, <span class="at">type =</span> <span class="dv">3</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Call:
lm(formula = weight ~ hindfoot_length + as.factor(sex), data = surveys_subset)

Residuals:
    Min      1Q  Median      3Q     Max 
-58.736 -12.860  -3.985   3.519 222.859 

Coefficients:
                 Estimate Std. Error t value Pr(&gt;|t|)    
(Intercept)     -31.55118    1.41051 -22.369  &lt; 2e-16 ***
hindfoot_length   2.57297    0.04414  58.287  &lt; 2e-16 ***
as.factor(sex)M  -3.64249    0.84599  -4.306 1.71e-05 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 26.15 on 3832 degrees of freedom
  (514 observations deleted due to missingness)
Multiple R-squared:  0.4702,    Adjusted R-squared:  0.4699 
F-statistic:  1700 on 2 and 3832 DF,  p-value: &lt; 2.2e-16</code></pre>
</div>
</div>
<p>Interactions between covariates are modeled by introducing additional covariates of the form <span class="math inline">\(x_i x_j\)</span>. <strong>An interaction occurs when an independent variable has a different effect on the outcome depending on the values of another independent variable. They are super important!</strong> To estimate the coefficients associated to an interaction, * is used in the call to <code>lm</code>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a>model <span class="ot">&lt;-</span> <span class="fu">lm</span>(weight<span class="sc">~</span>hindfoot_length<span class="sc">*</span><span class="fu">as.factor</span>(sex), <span class="at">data=</span>surveys_subset)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We interpret the coefficient of the interaction term as we do all other regression coefficients. Per unit change in <span class="math inline">\(x_i x_j\)</span> (here, the interaction between hindfoot length and sex), the associated regression coefficient measures the change in the response (weight).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a>surveys_subset <span class="sc">%&gt;%</span> <span class="fu">subset</span>(<span class="sc">!</span> <span class="fu">is.na</span>(sex)) <span class="sc">%&gt;%</span></span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> hindfoot_length, <span class="at">y =</span> weight, <span class="at">color =</span> <span class="fu">as.factor</span>(sex))) <span class="sc">+</span> <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_smooth</span>(<span class="at">method =</span> <span class="st">"lm"</span>, <span class="at">se =</span> F)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>`geom_smooth()` using formula = 'y ~ x'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Removed 285 rows containing non-finite values (`stat_smooth()`).</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Removed 285 rows containing missing values (`geom_point()`).</code></pre>
</div>
<div class="cell-output-display">
<p><img src="lec08-linear-models_files/figure-html/unnamed-chunk-7-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<section id="challenge-1" class="level4" data-number="9.4.2.1">
<h4 data-number="9.4.2.1" class="anchored" data-anchor-id="challenge-1"><span class="header-section-number">9.4.2.1</span> Challenge</h4>
<p>Is there a significant interaction between hindfoot length and sex on weight?</p>
</section>
</section>
<section id="the-relationship-between-linear-models-and-anova" class="level3" data-number="9.4.3">
<h3 data-number="9.4.3" class="anchored" data-anchor-id="the-relationship-between-linear-models-and-anova"><span class="header-section-number">9.4.3</span> The relationship between linear models and ANOVA</h3>
<p>Analysis of Variance (ANOVA) is a statistical technique that is used to analyze variation in observations within and between categories, and to test if two or more population means are equal. <strong>Importantly, ANOVA is mathematically equivalent to regression with covariates that are all categorical.</strong> This means the assumptions of linear regression apply when preforming an ANOVA.</p>
<p>ANOVA can be implemented in R as follows:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>model <span class="ot">&lt;-</span> <span class="fu">lm</span>(weight<span class="sc">~</span><span class="fu">as.factor</span>(sex), <span class="at">data=</span>surveys_subset)</span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a><span class="fu">anova</span>(model) <span class="co"># must wrap the anova() around a lm()/model</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Analysis of Variance Table

Response: weight
                 Df  Sum Sq Mean Sq F value Pr(&gt;F)
as.factor(sex)    1     730  730.48  0.5368 0.4638
Residuals      4014 5461759 1360.68               </code></pre>
</div>
</div>
<p>Briefly, ANOVA tests differences between means by <strong>decomposing the total sum of squares (variance in the observations) into the variance due to the level under investigation (e.g., sex) and its factors (M, F)</strong>. The <span class="math inline">\(F\)</span>-statistic for each level (excluding the residuals) is the mean square, divided by the residual mean square, and should be <span class="math inline">\(\sim 1\)</span> if the corresponding effects are <span class="math inline">\(=0\)</span>. We reject the null hypothesis (there is no differences in group means) for values of <span class="math inline">\(F\)</span> that are inconsistent with its distribution under the null hypothesis; recall from last time that the <span class="math inline">\(p\)</span>-value measures the probability of observing data more extreme than the calculated value of <span class="math inline">\(F\)</span> from the data, under the null hypothesis (i.e., the factors at some level have no effect). If <span class="math inline">\(p &lt; \alpha\)</span>, we reject the null hypothesis at significance level <span class="math inline">\(\alpha\)</span>.</p>
<section id="anova-with-two-or-more-covariates" class="level4" data-number="9.4.3.1">
<h4 data-number="9.4.3.1" class="anchored" data-anchor-id="anova-with-two-or-more-covariates"><span class="header-section-number">9.4.3.1</span> ANOVA with two or more covariates</h4>
<div class="cell">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a>model <span class="ot">&lt;-</span> <span class="fu">lm</span>(weight<span class="sc">~</span>hindfoot_length<span class="sc">*</span><span class="fu">as.factor</span>(sex), <span class="at">data=</span>surveys_subset)</span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a><span class="co"># wrap Anova(), not summary()</span></span>
<span id="cb30-4"><a href="#cb30-4" aria-hidden="true" tabindex="-1"></a><span class="fu">Anova</span>(model, <span class="at">type =</span> <span class="dv">3</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Anova Table (Type III tests)

Response: weight
                                Sum Sq   Df   F value  Pr(&gt;F)    
(Intercept)                     203901    1  298.3134 &lt; 2e-16 ***
hindfoot_length                1151284    1 1684.3645 &lt; 2e-16 ***
as.factor(sex)                      48    1    0.0709 0.79010    
hindfoot_length:as.factor(sex)    1951    1    2.8544 0.09121 .  
Residuals                      2618537 3831                      
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</code></pre>
</div>
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a><span class="co"># compare p-values with summary(), which are slightly different</span></span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a><span class="fu">anova</span>(model)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Analysis of Variance Table

Response: weight
                                 Df  Sum Sq Mean Sq   F value    Pr(&gt;F)    
hindfoot_length                   1 2312681 2312681 3383.5236 &lt; 2.2e-16 ***
as.factor(sex)                    1   12677   12677   18.5473 1.699e-05 ***
hindfoot_length:as.factor(sex)    1    1951    1951    2.8544   0.09121 .  
Residuals                      3831 2618537     684                        
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</code></pre>
</div>
</div>
<p>Instead of wrapping <code>summary()</code> around our linear model object, we use the <code>Anova()</code> function from the <code>cars</code> package to test the significance of our predictors. A Type 1 ANOVA (sequential sum of squares) will test the main effect of variable A, followed by the main effect of variable B after the main effect of A, followed by the interaction effect of AB. Since it tests factor A without controlling for factor B, the order that you specify your independent variables will matter! A Type 2 ANOVA (hierarchical sum of squares) tests for the main effect of variable A and variable B, but it does not assume that there is an interaction. If you have a factorial design, you should not be using a Type 2 ANOVA. Finally, a Type 3 ANOVA (marginal sum of squares) tests for the presence of a main effect while controlling for the other main effect and interaction; that is, it tests the effect of variable A if all other variables are already considered. This approach is therefore valid in the presence of significant interactions.</p>
<p>The type of ANOVA you use will matter when you have more than one independent variable, and especially if you have unbalanced data. By default, R uses type II sums of squares. Above, we are use Type 3 sums of squares as we care about the interaction between hindfoot length and sex, and we don’t prioritize the effects of one variable.</p>
</section>
</section>
<section id="a-fun-application" class="level3" data-number="9.4.4">
<h3 data-number="9.4.4" class="anchored" data-anchor-id="a-fun-application"><span class="header-section-number">9.4.4</span> A fun application!</h3>
<p>In quantitative genetics, regression is used to estimate the strength of directional, stabilizing, or disruptive selection on continuous traits (e.g., beak length) controlled by a large number of genes of small additive effect. Without getting into the weeds, the regression of relative fitness (or some proxy for fitness) on trait value provides an estimate of the selection <em>differential</em> <span class="math inline">\(S\)</span>, defined as the co-variance between fitness and the trait. The observation linear models could be used to estimate selection on on or more quantitative traits (possibly correlated) was operationalized in 1983 by <a href="https://onlinelibrary.wiley.com/doi/10.1111/j.1558-5646.1983.tb00236.x">Lande &amp; Arnold</a>. When trait measurements are normalized, the slope of the regression equals the strength of directional selection on the trait. Regression has been applied in a wide range of plant and animal taxa to estimate selection and, when there are multiple traits under investigation, the relative importance of direct vs indirect selection on evolution.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a>generate_data_directionalselection <span class="ot">&lt;-</span> <span class="cf">function</span>(<span class="at">n =</span> <span class="dv">100</span>, values){</span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb34-3"><a href="#cb34-3" aria-hidden="true" tabindex="-1"></a>  trait_data <span class="ot">&lt;-</span> <span class="fu">sort</span>(<span class="fu">sample</span>(<span class="at">size =</span> n, <span class="at">x =</span> values, <span class="at">replace =</span> T))</span>
<span id="cb34-4"><a href="#cb34-4" aria-hidden="true" tabindex="-1"></a>  fitness_data <span class="ot">&lt;-</span> <span class="dv">1</span> <span class="sc">-</span> <span class="fu">dexp</span>(trait_data, <span class="at">rate =</span> <span class="dv">1</span>) <span class="sc">+</span> <span class="fu">rnorm</span>(n, <span class="at">mean =</span> <span class="dv">0</span>, <span class="at">sd =</span> <span class="fl">0.05</span>)</span>
<span id="cb34-5"><a href="#cb34-5" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb34-6"><a href="#cb34-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(</span>
<span id="cb34-7"><a href="#cb34-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">data.frame</span>(</span>
<span id="cb34-8"><a href="#cb34-8" aria-hidden="true" tabindex="-1"></a>      <span class="at">trait =</span> trait_data, <span class="at">fitness =</span> fitness_data<span class="sc">/</span><span class="fu">mean</span>(fitness_data)</span>
<span id="cb34-9"><a href="#cb34-9" aria-hidden="true" tabindex="-1"></a>      )</span>
<span id="cb34-10"><a href="#cb34-10" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb34-11"><a href="#cb34-11" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb34-12"><a href="#cb34-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-13"><a href="#cb34-13" aria-hidden="true" tabindex="-1"></a>data <span class="ot">&lt;-</span> <span class="fu">generate_data_directionalselection</span>(<span class="at">values =</span> <span class="fu">seq</span>(<span class="dv">0</span>,<span class="dv">3</span>,<span class="fl">0.01</span>))</span>
<span id="cb34-14"><a href="#cb34-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-15"><a href="#cb34-15" aria-hidden="true" tabindex="-1"></a>model1 <span class="ot">&lt;-</span> <span class="fu">lm</span>(fitness<span class="sc">~</span><span class="fu">scale</span>(trait), data)</span>
<span id="cb34-16"><a href="#cb34-16" aria-hidden="true" tabindex="-1"></a><span class="co"># scale function transforms trait data so that mean=0, variance=1</span></span>
<span id="cb34-17"><a href="#cb34-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-18"><a href="#cb34-18" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(model1) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Call:
lm(formula = fitness ~ scale(trait), data = data)

Residuals:
     Min       1Q   Median       3Q      Max 
-0.36874 -0.11840  0.03626  0.13023  0.30442 

Coefficients:
             Estimate Std. Error t value Pr(&gt;|t|)    
(Intercept)   1.00000    0.01590   62.90   &lt;2e-16 ***
scale(trait)  0.37040    0.01598   23.18   &lt;2e-16 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 0.159 on 98 degrees of freedom
Multiple R-squared:  0.8457,    Adjusted R-squared:  0.8442 
F-statistic: 537.3 on 1 and 98 DF,  p-value: &lt; 2.2e-16</code></pre>
</div>
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a><span class="do">### fitness changed by ~50% per sd change in trait value!</span></span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb36-3"><a href="#cb36-3" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(data, <span class="fu">aes</span>(<span class="at">x =</span> <span class="fu">scale</span>(trait), <span class="at">y =</span> fitness)) <span class="sc">+</span> </span>
<span id="cb36-4"><a href="#cb36-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">color =</span> <span class="st">"gray"</span>) <span class="sc">+</span></span>
<span id="cb36-5"><a href="#cb36-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_smooth</span>(<span class="at">method =</span> <span class="st">"lm"</span>, <span class="at">formula =</span> y <span class="sc">~</span> x, <span class="at">se =</span> F) <span class="sc">+</span></span>
<span id="cb36-6"><a href="#cb36-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_smooth</span>(<span class="at">method =</span> <span class="st">"lm"</span>, <span class="at">formula =</span> y <span class="sc">~</span> x <span class="sc">+</span> <span class="fu">I</span>(x<span class="sc">^</span><span class="dv">2</span>), <span class="at">se =</span> F,</span>
<span id="cb36-7"><a href="#cb36-7" aria-hidden="true" tabindex="-1"></a>              <span class="at">color =</span> <span class="st">"red"</span>) <span class="sc">+</span></span>
<span id="cb36-8"><a href="#cb36-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">"standardized trait value"</span>, <span class="at">y =</span> <span class="st">"relative fitness"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="lec08-linear-models_files/figure-html/unnamed-chunk-10-1.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a>model2 <span class="ot">&lt;-</span> <span class="fu">lm</span>(fitness<span class="sc">~</span><span class="fu">scale</span>(trait)<span class="sc">+</span><span class="fu">scale</span>(trait<span class="sc">^</span><span class="dv">2</span>), data)</span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(model2)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Call:
lm(formula = fitness ~ scale(trait) + scale(trait^2), data = data)

Residuals:
      Min        1Q    Median        3Q       Max 
-0.157196 -0.059319  0.000927  0.042577  0.199743 

Coefficients:
                Estimate Std. Error t value Pr(&gt;|t|)    
(Intercept)     1.000000   0.007575  132.02   &lt;2e-16 ***
scale(trait)    0.885050   0.029141   30.37   &lt;2e-16 ***
scale(trait^2) -0.533167   0.029141  -18.30   &lt;2e-16 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 0.07575 on 97 degrees of freedom
Multiple R-squared:  0.9653,    Adjusted R-squared:  0.9646 
F-statistic:  1351 on 2 and 97 DF,  p-value: &lt; 2.2e-16</code></pre>
</div>
</div>
<p>Here, the blue (red) regression line is the linear (quadratic) individual selection surface, estimated by regressing fitness on trait value (and its square) simulated under a model of directional selection.</p>
<section id="other-applications" class="level4" data-number="9.4.4.1">
<h4 data-number="9.4.4.1" class="anchored" data-anchor-id="other-applications"><span class="header-section-number">9.4.4.1</span> Other applications…</h4>
<p>Regression has also been used in quantitative genetics to estimate the hertiability of traits (i.e., the proportion of variance in the trait that is explained by the additive action of genes), and to preform Genome-Wide Association Studies (GWAS) and uncover the genetic basis of complex traits. One form of GWAS involves regressing phenotype (e.g., disease status) on the existence/non-existence of genetic differences (most often, SNPs) among individuals sampled, and identifying those which have a significant effect on phenotype.</p>
</section>
</section>
</section>
<section id="generalized-linear-models-theory-and-examples" class="level2" data-number="9.5">
<h2 data-number="9.5" class="anchored" data-anchor-id="generalized-linear-models-theory-and-examples"><span class="header-section-number">9.5</span> Generalized linear models: theory and examples</h2>
<p>So far we have seen how regression can be used to fit linear models when the distribution of the data is approximately normal or when the data can be transformed so that this assumption is not violated. What if we were, say, interested in a <em>binary</em> response (e.g., disease status) and how it changes with a continuous predictor (e.g., age)? In this case, one can use a special kind of linear model called <em>logistic regression</em> to estimate the additive effect of predictor(s) on the binary response. <strong>Generalized linear models (GLMs) are useful when the response has a non-normal distribution, and transformation of the data is undesirable or impossible.</strong> A GLM takes the form</p>
<p><span class="math display">\[E(\boldsymbol{y}|\boldsymbol{X}) = g^{-1}(\boldsymbol{X} \boldsymbol{\beta}),\]</span></p>
<p>where <span class="math inline">\(g(\cdot)\)</span> is a smooth and invertible <em>link</em> function taking the conditional expectation on the LHS to the linear predictor on the RHS. The link function for distributions in the overdispersed exponential family (including the Normal, Gamma, Exponential, Poisson, and Multinomial) are known. GLMs with these data distributions can be implemented in R by calling <code>glm()</code> with the appropriate family specified:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a>result_binary <span class="ot">&lt;-</span> <span class="fu">glm</span>(<span class="fu">as.factor</span>(sex)<span class="sc">~</span>weight, </span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a>                     <span class="at">family=</span>binomial, </span>
<span id="cb39-3"><a href="#cb39-3" aria-hidden="true" tabindex="-1"></a>                     <span class="at">data=</span>surveys_subset)</span>
<span id="cb39-4"><a href="#cb39-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-5"><a href="#cb39-5" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(result_binary)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Call:
glm(formula = as.factor(sex) ~ weight, family = binomial, data = surveys_subset)

Deviance Residuals: 
   Min      1Q  Median      3Q     Max  
-1.222  -1.214   1.134   1.143   1.204  

Coefficients:
              Estimate Std. Error z value Pr(&gt;|z|)  
(Intercept)  0.1074131  0.0482240   2.227   0.0259 *
weight      -0.0006272  0.0008560  -0.733   0.4638  
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

(Dispersion parameter for binomial family taken to be 1)

    Null deviance: 5560.8  on 4015  degrees of freedom
Residual deviance: 5560.3  on 4014  degrees of freedom
  (333 observations deleted due to missingness)
AIC: 5564.3

Number of Fisher Scoring iterations: 3</code></pre>
</div>
</div>
<p>Above are estimates of coefficients of a regression of sex on weight, as well as some details about the procedure used to fit those parameters. All GLMs are fitted via maximum likelihood, using using numerical optimization methods like iteratively reweighted least squares.</p>
<p>Using the cannonical link function, logistic regression can be formulated as follows:</p>
<p><span class="math display">\[ \text{logit}(p) = \log\frac{p}{1-p} = \boldsymbol{X} \boldsymbol{\beta} \iff p = e^{\boldsymbol{X} \boldsymbol{\beta}} = e^{\beta_0 + \beta_1 x_1 + \dots + \beta_p x_p}. \]</span></p>
<p><span class="math inline">\(\text{logit}(p)\)</span> is called the <em>log-odds</em>, which can be thought of as a likelihood the response takes on the value one. Under logistic regression, the log-odds ratio is modeled as a linear combination in the predictors <span class="math inline">\((= \boldsymbol{X} \boldsymbol{\beta})\)</span>. Since the regression coefficients appear only through the logit-transformed proportions, their interpretation is somewhat different under logistic regression and other GLMs. <strong>Importantly, changing <span class="math inline">\(x_j\)</span> by one unit, all else constant, results in change <span class="math inline">\(\beta_j\)</span> to the link-transformed response. This is how the effect sizes are interpreted for GLMs like the one fitted above.</strong></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a>surveys_subset <span class="sc">%&gt;%</span> <span class="fu">filter</span>(<span class="sc">!</span> <span class="fu">is.na</span>(sex)) <span class="sc">%&gt;%</span></span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">ID =</span> <span class="fu">ifelse</span>(sex <span class="sc">==</span> <span class="st">"M"</span>, <span class="dv">0</span>, <span class="dv">1</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb41-3"><a href="#cb41-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> weight, <span class="at">y =</span> ID)) <span class="sc">+</span> <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb41-4"><a href="#cb41-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_smooth</span>(<span class="at">method =</span> <span class="st">"glm"</span>, </span>
<span id="cb41-5"><a href="#cb41-5" aria-hidden="true" tabindex="-1"></a>              <span class="at">method.args =</span> <span class="fu">list</span>(<span class="at">family =</span> <span class="st">"binomial"</span>),</span>
<span id="cb41-6"><a href="#cb41-6" aria-hidden="true" tabindex="-1"></a>              <span class="at">se =</span> T</span>
<span id="cb41-7"><a href="#cb41-7" aria-hidden="true" tabindex="-1"></a>              )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>`geom_smooth()` using formula = 'y ~ x'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Removed 104 rows containing non-finite values (`stat_smooth()`).</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Removed 104 rows containing missing values (`geom_point()`).</code></pre>
</div>
<div class="cell-output-display">
<p><img src="lec08-linear-models_files/figure-html/unnamed-chunk-11-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Here are some common types of response variables and their corresponding distributions:</p>
<ul>
<li><strong>Count data</strong> (positive integers only): <strong>Poisson</strong> distribution</li>
<li><strong>Over-dispersed count</strong> data (when the count data is more spread out than “expected”): <strong>negative binomial</strong> distribution</li>
<li><strong>Binary</strong> data (two discrete categories): <strong>binomial</strong> distribution</li>
</ul>
<section id="returning-to-the-disease-age-example" class="level3" data-number="9.5.1">
<h3 data-number="9.5.1" class="anchored" data-anchor-id="returning-to-the-disease-age-example"><span class="header-section-number">9.5.1</span> Returning to the disease-age example…</h3>
<p>Below we simulate binary disease data (0,1) for patients of several ages, assuming the the log-odds of disease is a linear function of age. We then fit a logistic regression to this data to determine the effect of age on disease risk. Next, we write a function to do a <strong>power analysis</strong>. That is, we determine the sample size that is required so that simulating the age-disease data repeatedly we identify a significant effect of age on disease status (i.e., reject the null hypothesis) at level <span class="math inline">\(\alpha = 0.01\)</span> <em>at least <span class="math inline">\(99\%\)</span> of the time</em>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a>data_generator <span class="ot">&lt;-</span> <span class="cf">function</span>(<span class="at">sample_size =</span> <span class="dv">100</span>){</span>
<span id="cb45-2"><a href="#cb45-2" aria-hidden="true" tabindex="-1"></a>  age <span class="ot">&lt;-</span> <span class="fu">sample</span>(<span class="at">size =</span> sample_size, <span class="at">x =</span> <span class="fu">seq</span>(<span class="dv">0</span>,<span class="dv">100</span>,<span class="fl">0.1</span>), <span class="at">replace =</span> T)</span>
<span id="cb45-3"><a href="#cb45-3" aria-hidden="true" tabindex="-1"></a>  linear_predictor <span class="ot">&lt;-</span> <span class="fl">0.8</span><span class="sc">*</span><span class="fu">scale</span>(age)</span>
<span id="cb45-4"><a href="#cb45-4" aria-hidden="true" tabindex="-1"></a>  prob <span class="ot">&lt;-</span> <span class="dv">1</span><span class="sc">/</span>(<span class="dv">1</span><span class="sc">+</span><span class="fu">exp</span>(<span class="sc">-</span>linear_predictor))</span>
<span id="cb45-5"><a href="#cb45-5" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb45-6"><a href="#cb45-6" aria-hidden="true" tabindex="-1"></a>  disease_status <span class="ot">&lt;-</span> <span class="fu">c</span>()</span>
<span id="cb45-7"><a href="#cb45-7" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb45-8"><a href="#cb45-8" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(prob)){</span>
<span id="cb45-9"><a href="#cb45-9" aria-hidden="true" tabindex="-1"></a>  disease_status[i] <span class="ot">&lt;-</span> <span class="fu">rbinom</span>(<span class="at">n =</span> <span class="dv">1</span>, <span class="at">size =</span> <span class="dv">1</span>, <span class="at">prob =</span> prob[i])</span>
<span id="cb45-10"><a href="#cb45-10" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb45-11"><a href="#cb45-11" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb45-12"><a href="#cb45-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(<span class="fu">data.frame</span>(<span class="at">age =</span> age, <span class="at">disease_status =</span> disease_status))</span>
<span id="cb45-13"><a href="#cb45-13" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb45-14"><a href="#cb45-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-15"><a href="#cb45-15" aria-hidden="true" tabindex="-1"></a>data <span class="ot">&lt;-</span> <span class="fu">data_generator</span>()</span>
<span id="cb45-16"><a href="#cb45-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-17"><a href="#cb45-17" aria-hidden="true" tabindex="-1"></a>data <span class="sc">%&gt;%</span> <span class="fu">pivot_longer</span>(<span class="sc">!</span> age) <span class="sc">%&gt;%</span> </span>
<span id="cb45-18"><a href="#cb45-18" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> age, <span class="at">y =</span> value)) <span class="sc">+</span> <span class="fu">geom_point</span>() <span class="sc">+</span> </span>
<span id="cb45-19"><a href="#cb45-19" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_smooth</span>(<span class="at">method =</span> <span class="st">"glm"</span>, <span class="at">method.args =</span> <span class="fu">list</span>(<span class="at">family =</span> <span class="st">"binomial"</span>)</span>
<span id="cb45-20"><a href="#cb45-20" aria-hidden="true" tabindex="-1"></a>              ) <span class="sc">+</span> <span class="fu">labs</span>(<span class="at">y =</span> <span class="st">"prob. of disease (i.e., disease status =1)"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>`geom_smooth()` using formula = 'y ~ x'</code></pre>
</div>
<div class="cell-output-display">
<p><img src="lec08-linear-models_files/figure-html/unnamed-chunk-12-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a>model <span class="ot">&lt;-</span> <span class="fu">glm</span>(disease_status<span class="sc">~</span><span class="fu">scale</span>(age), <span class="at">family =</span> binomial, <span class="at">data =</span> data)</span>
<span id="cb47-2"><a href="#cb47-2" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(model)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Call:
glm(formula = disease_status ~ scale(age), family = binomial, 
    data = data)

Deviance Residuals: 
    Min       1Q   Median       3Q      Max  
-1.6236  -1.0219  -0.7782   1.0535   1.6736  

Coefficients:
            Estimate Std. Error z value Pr(&gt;|z|)   
(Intercept) -0.08469    0.21035  -0.403  0.68722   
scale(age)   0.65288    0.21870   2.985  0.00283 **
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

(Dispersion parameter for binomial family taken to be 1)

    Null deviance: 138.47  on 99  degrees of freedom
Residual deviance: 128.78  on 98  degrees of freedom
AIC: 132.78

Number of Fisher Scoring iterations: 4</code></pre>
</div>
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a>power_analysis_function <span class="ot">&lt;-</span> <span class="cf">function</span>(sample_size){</span>
<span id="cb49-2"><a href="#cb49-2" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb49-3"><a href="#cb49-3" aria-hidden="true" tabindex="-1"></a>  sims <span class="ot">&lt;-</span> <span class="dv">1000</span></span>
<span id="cb49-4"><a href="#cb49-4" aria-hidden="true" tabindex="-1"></a>  pvalues <span class="ot">&lt;-</span> <span class="fu">c</span>()</span>
<span id="cb49-5"><a href="#cb49-5" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span>sims){</span>
<span id="cb49-6"><a href="#cb49-6" aria-hidden="true" tabindex="-1"></a>  data <span class="ot">&lt;-</span> <span class="fu">data_generator</span>(sample_size)</span>
<span id="cb49-7"><a href="#cb49-7" aria-hidden="true" tabindex="-1"></a>  model <span class="ot">&lt;-</span> <span class="fu">glm</span>(disease_status<span class="sc">~</span><span class="fu">scale</span>(age), <span class="at">family =</span> binomial, <span class="at">data =</span> data)</span>
<span id="cb49-8"><a href="#cb49-8" aria-hidden="true" tabindex="-1"></a>  pvalues[i] <span class="ot">&lt;-</span> <span class="fu">summary</span>(model)<span class="sc">$</span>coefficients[<span class="dv">2</span>,<span class="dv">4</span>]</span>
<span id="cb49-9"><a href="#cb49-9" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb49-10"><a href="#cb49-10" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb49-11"><a href="#cb49-11" aria-hidden="true" tabindex="-1"></a>  power_estimate <span class="ot">&lt;-</span> <span class="fu">length</span>(<span class="fu">which</span>(pvalues <span class="sc">&lt;</span> <span class="fl">0.01</span>))<span class="sc">/</span><span class="fu">length</span>(pvalues)</span>
<span id="cb49-12"><a href="#cb49-12" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb49-13"><a href="#cb49-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(power_estimate)</span>
<span id="cb49-14"><a href="#cb49-14" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb49-15"><a href="#cb49-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-16"><a href="#cb49-16" aria-hidden="true" tabindex="-1"></a>sample_sizes <span class="ot">&lt;-</span> <span class="fu">seq</span>(<span class="dv">100</span>,<span class="dv">200</span>,<span class="dv">10</span>); power_estimates <span class="ot">&lt;-</span> <span class="fu">c</span>()</span>
<span id="cb49-17"><a href="#cb49-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-18"><a href="#cb49-18" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(sample_sizes)){</span>
<span id="cb49-19"><a href="#cb49-19" aria-hidden="true" tabindex="-1"></a>  power_estimates[i] <span class="ot">&lt;-</span> <span class="fu">power_analysis_function</span>(<span class="at">sample_size =</span> sample_sizes[i])</span>
<span id="cb49-20"><a href="#cb49-20" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb49-21"><a href="#cb49-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-22"><a href="#cb49-22" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">kable</span>(<span class="fu">cbind</span>(<span class="at">n =</span> sample_sizes, <span class="at">power =</span> power_estimates))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="table table-sm table-striped small">
<thead>
<tr class="header">
<th style="text-align: right;">n</th>
<th style="text-align: right;">power</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: right;">100</td>
<td style="text-align: right;">0.895</td>
</tr>
<tr class="even">
<td style="text-align: right;">110</td>
<td style="text-align: right;">0.933</td>
</tr>
<tr class="odd">
<td style="text-align: right;">120</td>
<td style="text-align: right;">0.938</td>
</tr>
<tr class="even">
<td style="text-align: right;">130</td>
<td style="text-align: right;">0.956</td>
</tr>
<tr class="odd">
<td style="text-align: right;">140</td>
<td style="text-align: right;">0.971</td>
</tr>
<tr class="even">
<td style="text-align: right;">150</td>
<td style="text-align: right;">0.989</td>
</tr>
<tr class="odd">
<td style="text-align: right;">160</td>
<td style="text-align: right;">0.986</td>
</tr>
<tr class="even">
<td style="text-align: right;">170</td>
<td style="text-align: right;">0.991</td>
</tr>
<tr class="odd">
<td style="text-align: right;">180</td>
<td style="text-align: right;">0.997</td>
</tr>
<tr class="even">
<td style="text-align: right;">190</td>
<td style="text-align: right;">1.000</td>
</tr>
<tr class="odd">
<td style="text-align: right;">200</td>
<td style="text-align: right;">0.997</td>
</tr>
</tbody>
</table>
</div>
</div>


</section>
</section>
<section id="footnotes" class="footnotes footnotes-end-of-document" role="doc-endnotes">
<hr>
<ol>
<li id="fn1"><p>In R, the function <code>t()</code> can be used to transpose a vector or matrix.<a href="#fnref1" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
</ol>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="./lec07-intro-to-stats.html" class="pagination-link">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Introduction to inference</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./projects.html" class="pagination-link">
        <span class="nav-page-text"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Project description</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->



</body></html>